[Joshua Chiang] iP

ChickenChiang opened the pull request:
# Cluck! 🐔 
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

**Cluck!** helps you keep track of your tasks, so you can go about your day by putting all your eggs in one basket.
- text-based
- easy to use
- fun and interactive
- *Faster than a chicken can cross the road!* :

### All you need to do is:
1. Download it [here](https://github.com/ChickenChiang/ip/) 
2. Start running Cluck!
3. Add your tasks
4. Let Cluck! handle the rest

Did we mention it was ***FREE?!***

### Features:
- [x] Managing Task
- [x] Find Tasks
- [ ] Sort by date (coming soon!)

If you **Java programmer**, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        new Cluck("Your_File_Directory_Here").run();
    }
}
```

Commit:
Add Gradle support

ChickenChiang committed:
git tag Level-1

ChickenChiang committed:
Greet, Echo, Exit

Implement a skeletal version of Cluck that starts by greeting the user, simply echos commands entered by the user, and exits when the user types bye.

ChickenChiang committed:
Add, List

ChickenChiang committed:
Merge remote-tracking branch 'origin/master'

ChickenChiang committed:
Update .gitignore

ChickenChiang committed:
Now you can mark tasks as done and change them back to not done

ChickenChiang committed:
Merge branch 'master' of https://github.com/ChickenChiang/ip

ChickenChiang committed:
Added the following types of tasks: 1. Todos, which only have description, 2. Deadlines with description and due date, 3. Events with description, start and end time

ChickenChiang committed:
A-TextUiTesting

Use the input/output redirection technique to semi-automate the testing of Duke(or rather, CLUCK)

ChickenChiang committed:
A-Exceptions

Adding exception classes to handle errors by Duke

ChickenChiang committed:
Level-6 Delete and A-Collections

Now you can delete tasks from the todo list with Cluck. Tasks are also now stored in an ArrayList.

Jun-How commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,194 @@
+    static private final String MAKE_EVENT = "event";
+    static private final String DUE_DATE_FLAG = "/by ";
+    static private final String EVENT_START_FLAG = "/from ";
+    static private final String EVENT_END_FLAG = "/to ";
```
Should the "static" keyword come before the access modifier ("private" in this case)?
```private static final``` might be the preferred format.

Namings align with the coding standard, which is great!

Jun-How commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,15 @@
+    public String toString() {
+        return this.isMarked
+                ? String.format("[D][X] %1$s (by: %2$s)", this.description, this.dueDate)
+                : String.format("[D][ ] %1$s (by: %2$s)", this.description, this.dueDate);
```
Could this method make use of abstraction so that the ```toString()``` method in the parent class Task could be used here? Then, you would only need to add the correct task icon ("D" for Deadline here) and the due date, making it easier to read.

Jun-How commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,194 @@
+        System.out.println("    ____________________________________________________________");
+
+        boolean loop = true;
+        List<Task> toDoList = new ArrayList<>();
```
Perhaps a better name could be used for the ```boolean``` name sound more like a boolean. toDoList could also be renamed to be more specific and descriptive of a container of multiple tasks.
```suggestion
        boolean isActive = true;
        List<Task> tasks = new ArrayList<>();
```

Jun-How commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,17 @@
+    public String toString(){
+        return this.isMarked
+                ? String.format("[E][X] %1$s (from:%2$s to:%3$s)", this.description, this.startTime, this.endTime)
+                : String.format("[E][ ] %1$s (from:%2$s to:%3$s)", this.description, this.startTime, this.endTime);
```
Similar to the comment under Deadline, perhaps this could make use of the ```toString()``` method in the parent class Task?

Jun-How commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,194 @@
+                + "|  ___|| |     _    _  |  ___| | |  _\n"
+                + "| |    | |    | |  | | | |     | |/ /\n"
+                + "| |___ | |___ | |__| | | |___  |    \\\n"
+                + "|_____||_____||______| |_____| |_| \\_\\\n";
```
I like that you put in the extra effort to make a personalised logo :D

Jun-How commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,194 @@
+                    System.out.println("    ____________________________________________________________");
+                    System.out.println("    You gotta give me a command!");
+                    System.out.println("    ____________________________________________________________");
+
```
Perhaps remove unnecessary white spaces between lines of code, such as within the switch statement? So that the format is standardised. This applies to some parts above too.

Jun-How reviewed (changes requested):
In terms of coding standard, it looks mostly fine, except a few naming issues and unnecessary white spaces.
Some improvements requested below.

adam07018 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,194 @@
+    static private final String EVENT_END_FLAG = "/to ";
+
+    public static boolean isNumeric(String strNum) {
+        if (strNum == null) {
```
Love your code! Tidy and clean

adam07018 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,194 @@
+                        Integer itemNumber = Integer.parseInt(words[1]);
+                        if (itemNumber > toDoList.size() || itemNumber <= 0) {
+                            System.out.println("    ____________________________________________________________");
+                            System.out.println("    That's not...? In the list...? Buh caw?");
```
Interesting and Creative ui design

adam07018 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,194 @@
+        Scanner sc = new Scanner(System.in);
+
+        while (loop) {
+            String input = sc.nextLine();
```
Should we add a sc.close() at the end

adam07018 reviewed (commented):
Overall code is clean and nice

ChickenChiang committed:
Made Task class abstract and implemented makeSaveFormat() function

Making the Task class abstract ensures that it cannot be instantiated.

The new makeSaveFormat() function is an abstract method in Task, and it is overridden in each child class such that it returns a string that is formatted with respect to the child class for saving tasks.

ChickenChiang committed:
Added save functionality

Now the tasks in the todo list are saved into a txt file with consistent formmating with respect to the type of task.
It also captures the status of the task, whether or not it is marked or unmarked.

ChickenChiang committed:
Level-7. Save

Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when Cluck starts up.

ChickenChiang committed:
Level-8. Added functionality for deadlines to understand date and time and restructure of package

Instead of saving the due dates of deadlines as strings they are now LocalDateTime objects which uses the java.time.LocalDateTime package

ChickenChiang committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

ChickenChiang committed:
Using Gradle to automate building and testing for Cluck

Added support for gradle to the project as well as denpendencies and tests

ChickenChiang committed:
Changes to checkstyle and addition of JUnit test classes

ChickenChiang committed:
Merge branch 'master' into branch-level-7

ChickenChiang committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Tasks/Task.java

ChickenChiang committed:
Merge commit 'be7a1ad6d4dc6d185298f52a006cb21c7b77cf49'

ChickenChiang committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Tasks/Task.java

ChickenChiang committed:
Merge branch 'master' of https://github.com/ChickenChiang/ip

# Conflicts:
#	src/main/java/Tasks/Task.java

ChickenChiang committed:
Merge branch 'A-Exceptions'

Added Exceptions for better exception handling

ChickenChiang committed:
Reformatting for files and adding messages as a class

To allow for better OOP and more packages down the line, the files have been renamed and restructured.

ChickenChiang committed:
Replaced text with strings from Messages package

ChickenChiang committed:
Created TaskList class and its related exceptions

TaskList is created to handle task operations.

TaskIndexOutOfBoundsException is thrown when a given index is out of the range of the taskList
TaskNotFoundException is thrown when a task is not found within a taskList (function that throws this is yet to be implemented)

ChickenChiang committed:
Restructure of file to fit OOP standard

TaskList related functions are moved from Cluck to TaskList class instead.
Now functions such as populating tasklist by reading saved file are under TaskList

Similarly Task related functions are moved from Cluck to Task class instead.

ChickenChiang committed:
Added functions for UI

Added greetUser and farewellUser functions for Ui to display start and end messages for Cluck, using the messages package

ChickenChiang committed:
Created commands package and Command class

Command class is a abstract class from which all commands will extend from.

ChickenChiang committed:
Created parser package and Parser class

Parser class will take user input in string format and convert it into its respective executable commands

ChickenChiang committed:
Command abstract class added to commands package

Command class is the parent class of all command classes, and contains abstract method execute().

ChickenChiang committed:
Changed Command class into and interface instead

It seems to make more sense for Command to be an interface that commands implement

ChickenChiang committed:
Refactoring of TaskList and Ui class to their respective packages

ChickenChiang committed:
Cluck class now concrete

An instantiated Cluck class now contains a Ui instance and a TaskList instance.
Also added start() and stop() functions to start and stop the Cluck instance

ChickenChiang committed:
Parser class now builds commands instead of Cluck

ChickenChiang committed:
Minor fix to Ui class

Ui class' readCommand() now waits for user input by using Scanner.nextLine()

ChickenChiang committed:
Minor fixes to Command interface and MissingArgumentException

Now the interface method execute() takes in a TaskList and returns a String containing the response

ChickenChiang committed:
Storage class added

Storage class added under package cluck.storage; Storage class will handle the reading and writing of saved files for cluck

ChickenChiang committed:
Storage class improvements

readSave() and writeSave() are now the only functions in storage. buildTaskFromSave() is now under Task class instead (to better follow OOP standards).
TaskList also now supports its own toSave() function so as to not violate OOP standards.

ChickenChiang committed:
Reformatted cluck to fit OOP principles

Minor fixes to cluck and reformatted to improve readability and abstraction.

ChickenChiang committed:
Added toString() method for TaskList class

TaskList now can return its tasks in String format

ChickenChiang committed:
Created ExitCommand and ListCommand classes

ExitCommand and ListCommand both implements the Command interface. Parser has also be edited to return ExitCommand or ListCommand based on user input

ChickenChiang committed:
MarkTaskCommand, UnmarkTaskCommand, ToDoCommand added

Minor fixes between files as well.
Now Parser creates the commands added based on switch case.

ChickenChiang committed:
Additional Commands created

DeadlineCommand, DeleteTaskCommand and EventCommand have been created and Parser now uses them to create commands

ChickenChiang committed:
Invalid Command and IsNotNumeric

Invalid Command added for when user input does not match any known command.

IsNumeric is replaced by IsNotNumeric for better readability

ChickenChiang committed:
Minor fixes

Cluck is now running as intended

ChickenChiang committed:
Merge branch 'A-MoreOOP'

ChickenChiang committed:
Minor fixes

Fields in instance of Cluck classare now final.
Added @throw for commandFactory function in Parser class.

ChickenChiang committed:
Added test files for gradle and minor fixes

ChickenChiang committed:
Added serach by keyword method for TaskList and Task

containsKeyword() function in Task class returns a boolean value, true if the description of the taks contains the keywrod given in the argument.

findMatches function in TaskList class iterates through its tasks and calls containsKeywords() on the task. If true, it is added to a temp list. The function findMatches then returns a string format of all the tasks that contains the keyword, stored in the temp list

ChickenChiang committed:
Created and Implemented FindTaskCommand

FindTaskCommand contains one field which is the key wrod the user is looking for. It then calls the findMatches() method on the Cluck task list and puts the key word as an argument.
As of now it is not case-sensitive.

ChickenChiang committed:
Modified find Command to be not case-sensitive.

This is done by converting both the keyword and description in to lower case.

ChickenChiang committed:
Added JUnit test for task checking

Tests if savedTasks is properly converted to Tasks by buildTaskFromSave Method in Task class.
Currently only tests on todo.

ChickenChiang committed:
Added exitCommandTest for Parser

Tests if given user input "bye", commandFactory produces an ExitCommand.

ChickenChiang committed:
Added JavaDocs

JavaDocs added to the commited class files.

ChickenChiang committed:
Added JavDocs for TaskList class

ChickenChiang committed:
Merge branch 'branch-Level-9' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/cluck/taskList/TaskList.java

ChickenChiang committed:
JavaDoc for Storage class added

ChickenChiang committed:
JavaDocs for Parser class added

ChickenChiang committed:
Minor fix for Messages

Removed excessive whitespaces in welcome message.

ChickenChiang committed:
Merge branch 'master' of https://github.com/ChickenChiang/ip into branch-A-JavaDoc

ChickenChiang committed:
JavaDocs for classes in commands package and exceptions package

ChickenChiang committed:
Rename taskList package to tasklist

ChickenChiang committed:
Merge branch 'branch-A-JavaDoc' into branch-A-CheckStyle

# Conflicts:
#	src/main/java/cluck/commands/EventCommand.java

ChickenChiang committed:
Merge branch 'branch-A-CheckStyle'

ChickenChiang committed:
Merge branch 'branch-A-JUnit'

* branch-A-JUnit:
  Added exitCommandTest for Parser
  Added JUnit test for task checking

# Conflicts:
#	src/main/java/cluck/commands/ExitCommand.java
#	src/main/java/cluck/tasks/Task.java
#	src/test/java/cluck/tasks/TaskTest.java

ChickenChiang committed:
Added dependencies for JavaFX to Gradle

ChickenChiang committed:
Fix Exit command and merge conflicts in Task class

ChickenChiang committed:
Make Cluck class extend JavaFx's Application class and created Launcher class

Cluck class also has a new constructor with no arguments for JavaFx to work properly.

ChickenChiang committed:
Creating a GUI for Duke

Follows everything up to tutorial 2 of [](https://se-education.org/guides/tutorials/javaFxPart2.html)

ChickenChiang committed:
Created DialogBox class for custom controls

ChickenChiang committed:
Implemented features for Cluck's dialog box to be interactable

- added sendButton.setOnMouseClicked(event) and userInput.setOnAction(event) to trigger handleUserInput() function
- getDialogLabel() function adds text label beside image of dialog box.
- handleUserInput() function controls actions taken after user input.

ChickenChiang committed:
Add custom behavior to DialogBox

Cluck's dialog box is now on the left side of the GUI while the user is on the right. This is done by implementing a flip() function in the custom dialog box class

ChickenChiang committed:
Added XML files and restructured the code for better  OOP

Classes that make up the Ui are now packages in the Ui package, namely DialogBox and MainWindow

XML files have also been added such that scenebuilder can be used to format JavaFx objects

ChickenChiang committed:
CheckStyle fixes

ChickenChiang committed:
Merge branch 'branch-Level-10-redo'

ChickenChiang committed:
Currently the jar file created with gradle using the shadow plugin is named 'duke.jar'
hich is not the
Jar file now correctly named

ChickenChiang committed:
Currently the current  jar file created using the shadow plugin in gradle is named "duke.jar"

Eventually the final products name will is "Cluck!"

This is to give it more personality and make it more user-friendly

To acheive this we simply edited the "build.gradle" file and change it such that  'archiveBaseName = Cluck!' under 'shadowJar'.

ChickenChiang committed:
Created CorruptedDataException

Currently if the saved file contains a task that is saved in the wrong format, cluck handles it by printing out a statement and the saved task.

To better handle such cases I added a CorruptedDataException which extends cluck exception. Here are the sequence of events after the changes:
1) task build fails in buildTaskFromSave() method in Task class,
2) exceptions are caught
3) CorruptedDataException is thrown containing the error statement and the incorrectly formatted save

ChickenChiang committed:
Fix Task class errors

There was an additional date time formatter left in the Task class, perhaps a vestige conflict merges.
Now there is only one protected static final DateTimeFormatter which all task related child-classes refer to

Changes were also made to the date time format of the formatter,  changed from "dd-MM-yyyy HH:mm" to "dd-MMM-yy HHmm"

ChickenChiang committed:
Fixes to Storage, Deadline, and Event classes

Minor fixes to ensure that Cluck operates normally.

Storage now has a try and catch clause to catch any exceptions when attempting to build a task from save. it catches any cluck exceptions and displays the error message to the user

ChickenChiang committed:
Merge pull request #2 from ChickenChiang/branch-DateTime-For-Events

Branch date time for events

ChickenChiang committed:
Cluck! now working in GUI

Previously Cluck!'s GUI merely repeated the user input, now it has been fixed such that Cluck! actually takes in the user input and returns a response after executing the command.

This was done by changing the getResponse method in Cluck class which takes in the user input as a String argument and parses it into a command, which is then executed and returns the response from executing the command.

ChickenChiang committed:
Merge pull request #3 from ChickenChiang/branch-functional-Cluck-GUI

Cluck! now working in GUI

ChickenChiang committed:
Edit to improve code quality in Cluck class

Previously Cluck contained high levels of nesting.

Additional methods have been added to improve readability and overall quality of code in the run() method in cluck.
getResponse() is called to get the response from a userInput. getResponse calls two other methods:
getCommand - takes in user input and parses it into a command
executeCommand - takes in a command and executes it, returning the response after executing the command

ChickenChiang committed:
Fixed accidental removal of import statement

ChickenChiang committed:
Merge branch 'master' into branch-A-CodeQuality

ChickenChiang committed:
Merge pull request #4 from ChickenChiang/branch-A-CodeQuality

Branch a code quality

ChickenChiang committed:
Merge branch 'master' of https://github.com/ChickenChiang/ip

ChickenChiang committed:
Currently the current  jar file created using the shadow plugin in gradle is named "duke.jar"

Eventually the final products name will is "Cluck!"

This is to give it more personality and make it more user-friendly

To acheive this we simply edited the "build.gradle" file and change it such that  'archiveBaseName = Cluck!' under 'shadowJar'.

ChickenChiang committed:
Merge pull request #5 from ChickenChiang/branch-A-JAR

Branch a jar

ChickenChiang committed:
Added assertions in private methods for Cluck

Ensure that response obtained after calling getResponse() within run() method is not null.
Ensure that getCommand method in Cluck should not have nuill as an argument.

ChickenChiang committed:
Merge pull request #6 from ChickenChiang/branch-A-Assertions

Added assertions in private methods for Cluck

ChickenChiang committed:
Upload Ui.png into docs

ChickenChiang committed:
Merge pull request #7 from ChickenChiang/branch-A-UserGuide

Upload Ui.png into docs

ChickenChiang committed:
Updated user guide for Cluck!

ChickenChiang committed:
Merge pull request #8 from ChickenChiang/branch-A-UserGuide

Updated user guide for Cluck!

ChickenChiang committed:
Update README.md

ChickenChiang committed:
Update README.md

ChickenChiang committed:
Restructured Parser class to fit code quality standards

Previously, the commandFactory method was far too big, and contains too many clauses. This makes looking at the code (much less bug fixing or tracing code) stroke-inducing. Further more, as more possible commands may be added in the future, the commandFactory method will become unbearably long.

To fix this, private methods to build commands for each case have been added. This also ensures that the Same Level of Abstraction(SLAP) is achieved within commandFactory() method.
The methods are named with the simple convention: "build<command_type>", and are pretty self-explainatory in their names. Simple commands such as ExitCommand, ListCommand, and InvalidCommands do not have their own build<command_type> methods as they do not require it.

ChickenChiang committed:
Removed Ui class as it was made redundant with JavaFX

ChickenChiang committed:
Giving Cluck! more personality

When user enters nothing in the command line, cluck now response with a joke. Try it out!

ChickenChiang committed:
NoCommand created

NoCommand is a place-holder command for when user does enters an empty string

ChickenChiang committed:
Merge pull request #9 from ChickenChiang/branch-A-CodeQuality2

Branch a code quality2

ChickenChiang committed:
Release for version v0.2 for Cluck!

ChickenChiang committed:
Refactor code for saving data

Currently Cluck! only writes in the svae file when the exit command is given, now the saved file is updated for any command that induces a change in the task list.

ChickenChiang committed:
Added flexibility for user date tiem inputs

Previously the date time input of the user has to be precise, for example if the user wants a deadline in march they would have to input "Mar" exactly.

This commit changes it such that the date time formatter is no longer case sensitive, and the user can even enter the numerical equivalent of the month they desire.

ChickenChiang committed:
Fixed userGuide and readme for github pages

ChickenChiang committed:
Merge pull request #10 from ChickenChiang/branch-fixUserGuide

Fixed userGuide and readme for github pages

ChickenChiang committed:
Rename user guide file to README.md so github pages is working

I didn't know it had to be a README.md lmao

ChickenChiang committed:
Merge branch 'master' of https://github.com/ChickenChiang/ip into branch-fixUserGuide

ChickenChiang committed:
Merge pull request #11 from ChickenChiang/branch-fixUserGuide

Branch fix user guide

ChickenChiang committed:
Edit readme for better formatting

ChickenChiang committed:
Merge pull request #12 from ChickenChiang/branch-fixUserGuide

Edit readme for better formatting

ChickenChiang committed:
Fixing anchoring for relative links in table of content

ChickenChiang committed:
Fixing anchor links and clarifying guide

ChickenChiang committed:
Merge pull request #13 from ChickenChiang/branch-fixUserGuide

Branch fix user guide

ChickenChiang committed:
Improve formatting of user guide

ChickenChiang committed:
Merge pull request #14 from ChickenChiang/branch-fixUserGuide

Improve formatting of user guide

ChickenChiang committed:
Fixing indentation levels

ChickenChiang committed:
Merge pull request #15 from ChickenChiang/branch-fixUserGuide

Fixing indentation levels

ChickenChiang committed:
Fixing indentation AGAIN

ChickenChiang committed:
Merge branch 'master' of https://github.com/ChickenChiang/ip into branch-fixUserGuide

ChickenChiang committed:
Merge pull request #16 from ChickenChiang/branch-fixUserGuide

Branch fix user guide

ChickenChiang committed:
Fixed pathing issue for saved file

The Cluck!.jar now creates a directory and a text file that saves the tasks of the user

ChickenChiang committed:
Minor codestyle fixes in JUnit test classes

ChickenChiang committed:
Merge pull request #18 from ChickenChiang/branch-FixSave

Fixed saving issue for Cluck!

ChickenChiang committed:
Merge branch 'master' into branch-C-NaturalDates

# Conflicts:
#	src/main/java/cluck/Cluck.java
#	src/main/java/cluck/storage/Storage.java

ChickenChiang committed:
Allow for more flexibility when entering dates

ChickenChiang committed:
no message

ChickenChiang committed:
Make Cluck! close on bye command

ChickenChiang committed:
Merge pull request #19 from ChickenChiang/branch-ExitGuiOnByeCommand

Make Cluck! close on bye command

ChickenChiang committed:
Merge branch 'master' into branch-C-NaturalDates

ChickenChiang committed:
Fixed date time formatting for tasks

Previous implementation to allow for more flexble parsing of date time given by the user cause some problems. In the previous commit I added optional patterns to allow for user to input various forms of date time, e.g. the user can input either 06, Jun, or June for a date on june. The user also has the option to omit the century of the year (i.e. the "20" in "2023") or minute of the time.

The change in FORMATTER in the task class cause the toString() and makeSaveFormat() methods in Event and Deadline classes to format the LocalDateTime classes into the wrong format, as the FORMATTER's optional patterns are all used to format the LocalDateTime, resulting in strange date strings like "21 03MarMarc 202323 1200".

To resolve the issue, I added an additional formatter by the name of default formatter where the DateTimeFormatter does not have optional patterns, and thus gives us consistent and correct string formats of the date when used by toString() and makeSaveFormat() methods.

ChickenChiang committed:
Merge pull request #20 from ChickenChiang/branch-C-NaturalDates

Branch c natural dates

ChickenChiang committed:
Updated user guide

User guide now contains additional information about the newly added date time format

ChickenChiang committed:
Merge pull request #21 from ChickenChiang/branch-UpdateUserGuide

Updated user guide

ChickenChiang committed:
Fix user guide formattiing

ChickenChiang committed:
Merge pull request #22 from ChickenChiang/branch-UpdateUserGuide

Fix user guide formattiing

ChickenChiang committed:
Delete Cluck!.zip

ChickenChiang committed:
Remove DateTimeHandler class

This class was a feature that was in progress, not to be committed in master

ChickenChiang committed:
Fix checkstyle errors

ChickenChiang committed:
fix checkstyle

ChickenChiang committed:
Merge branch 'master' into branch-C-NaturalDates

ChickenChiang committed:
Merge pull request #23 from ChickenChiang/branch-C-NaturalDates

Branch c natural dates
